quilt/tools/build.py

```diff
@@ -105,3 +118,95 @@ def build_package(username, package, yaml_path):
         _build_table(build_dir, store, '', tables)
         if readme is not None:
             _build_file(build_dir, store, 'README', rel_path=readme)
+
+def splitext_no_dot(filename):
+    name, ext = os.path.splitext(filename)
+    ext.strip('.')
+    return name, ext.strip('.')
+
+def generate_build_file(startpath, outfilename='build.yml'):
+    startbase = os.path.basename(startpath)
+    buildfiles = {startbase : {}}
+    buildtables = {startbase : {}}
+
+    def add_to_buildfiles(path, files):
+        try:
+            safepath = [_pythonize_name(d) if d != '.' else '.' for d in path]
+        except BuildException:
+            warning = "Warning: could not determine a Python-legal name for {path}; skipping."
+            print(warning.format(path=os.sep.join(path)))
+            return
+
+        ptr = buildfiles
+        for folder in path:
+            ptr = ptr.setdefault(folder, {})
+
+        for file in files:
+            fullpath = os.path.join(*safepath, file)
+            name, ext = splitext_no_dot(file)
+            ptr[_pythonize_name(name)] = fullpath
+
+    def add_to_buildtables(path, files):
+        try:
+            safepath = [_pythonize_name(d) if d != '.' else '.' for d in path]
+        except BuildException:
+            warning = "Warning: could not determine a Python-legal name for {path}; skipping."
+            print(warning.format(path=os.sep.join(path)))
+            return
+
+        ptr = buildtables
+        for folder in safepath:
+            ptr = ptr.setdefault(folder, {})
+        for file in files:
+            fullpath = "/".join(path + [file])
+            name, ext = splitext_no_dot(file)
+            # pythonize name
+            ptr[_pythonize_name(name)] = [ext.lower(), fullpath]
+
+    for root, dirs, files in os.walk(startpath):
+        # skip hidden directories
+        for d in dirs:
+            if d.startswith('.'):
+                dirs.remove(d)
+
+        rel_path = os.path.relpath(root, startpath)
+        path = rel_path.split(os.sep)
+
+        tablefiles = []
+        rawfiles = []
+        for file in files:
+            # skip hidden files
+            if file.startswith('.'):
+                continue
+            try:
+                name, ext = splitext_no_dot(file)
+                # separate files into tables and raw
+                if ext.lower() in TARGET['pandas']:
+                    tablefiles.append(file)
+                else:
+                    rawfiles.append(file)
+            except ValueError:
+                # File with no extension
+                rawfiles.append(file)
+
+        if rawfiles:
+            add_to_buildfiles(path, rawfiles)
```

Comment:
I think the logic would be simpler if `add_to_buildfiles`/`add_to_buildtables` took a single file instead of a list of files. That is, I don't see much benefit in building a list of files, then having the helper functions loop over the files.

I guess that would mean calculating `safepath` lots of times... But I'd move it out of the helper functions and put it after `path = rel_path.split(os.sep)`. Then just pass a joined safe path into the helpers.

In fact, even the `ptr = ptr.setdefault(folder, {})` logic could be outside the helper. When doing `os.walk`, you know exactly when you need to create the nested dict.

Anyways, just suggestions - don't need to rewrite everything if you don't agree.